### PR TITLE
[BACKPORT] Improve logging tests isolation

### PR DESCRIPTION
**Pull request by taburet**
Provides a new JUnit test rule for use by logging related tests.
The rule (IsolatedLoggingRule) fully isolates the logging state
from other tests.

Backport of https://github.com/hazelcast/hazelcast/pull/12714

### CODE REVIEW

**Comment by Holmistr** [comment]
run-lab-run

File: hazelcast/src/test/java/com/hazelcast/test/IsolatedLoggingRule.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+import java.lang.reflect.Field;
+
+/**
+ * Fully isolates the shared global logging state for a test. Also provides
+ * utilities which are useful while testing logging related things.
+ * <p>
+ * Make sure you are using a serial test runner if you use this rule.
```
Comment (blazember) [inline hazelcast/src/test/java/com/hazelcast/test/IsolatedLoggingRule.java]: Same idea as in the mainstream PR:


>I think we can even enforce it. `Description` param of `apply` seems to know about the test and the test class being executed. 
>
>Adding something like the following to the beginning of the `apply` method below may do the trick.
>```
>// making sure that there are no concurrent test executions so we don't impact their results unintentionally
>RunWith runWith = description.getTestClass().getAnnotation(RunWith.class);
>if (!HazelcastSerialClassRunner.class.equals(runWith.value())) {
>  throw new IllegalStateException("Tests using this rule must use serial runner");
>}
>```

**Comment by Donnerbart** [comment]
run-lab-run

**Comment by taburet** [comment]
run-lab-run

**Pull request merged**